snorkel/map/core.py

```diff
@@ -28,19 +33,90 @@ def get_parameters(
     return params[0]
 
 
+def is_hashable(obj: Any) -> bool:
+    """Test if object is hashable via duck typing.
+
+    NB: not using `collections.Hashable` as some objects
+    (e.g. pandas.Series) have a `__hash__` method to throw
+    a more specific exception.
+    """
+    try:
+        hash(obj)
+        return True
+    except Exception:
+        return False
+
+
+def get_hashable(obj: Any) -> Hashable:
+    """Get a hashable version of a potentially unhashable object.
+
+    This helper is used for caching mapper outputs of data points.
+    For common data point formats (e.g. SimpleNamespace, pandas.Series),
+    produces hashable representations of the values using a `frozenset`.
+    For objects like `pandas.Series`, the name/index indentifier is dropped.
+
+    Parameters
+    ----------
+    obj
+        Object to get hashable version of
+
+    Returns
+    -------
+    Hashable
+        Hashable representation of object values
+
+    Raises
+    ------
+    ValueError
+        No hashable proxy for object
+    """
+    # If hashable already, just return
+    if is_hashable(obj):
+        return obj
+    # Get dictionary from SimpleNamespace
+    if isinstance(obj, SimpleNamespace):
+        obj = vars(obj)
+    # For dictionaries or pd.Series, construct a frozenset from items
+    # Also recurse on values in case they aren't hashable
+    if isinstance(obj, (dict, pd.Series)):
+        return frozenset((k, get_hashable(v)) for k, v in obj.items())
+    # For lists, recurse on values
+    if isinstance(obj, (list, tuple)):
+        return tuple(get_hashable(v) for v in obj)
+    # For NumPy arrays, hash the byte representation of the data array
+    if isinstance(obj, np.ndarray):
+        return obj.data.tobytes()
+    raise ValueError(f"Object {obj} has no hashing proxy.")
+
+
 class BaseMapper:
     """Base class for `Mapper` and `LambdaMapper`.
 
-    Implements mode setting and deep copy functionality.
+    Implements mode setting, memoization, and deep copy functionality.
 
     Raises
     ------
     NotImplementedError
         Subclasses need to implement `_generate_mapped_data_point`
     """
 
-    def _generate_mapped_data_point(self, x: DataPoint) -> DataPoint:
-        raise NotImplementedError
+    def __init__(self, memoize: bool) -> None:
+        self.memoize_outputs(memoize)
+        self.reset_cache()
+
+    def memoize_outputs(self, mode: bool) -> None:
+        """Set memoization mode.
+
+        Parameters
+        ----------
+        mode
+            Memoize mapper outputs?
+        """
+        self._memoize_outputs = mode
```

Comment:
Is this a little overkill?  Can't we just set the `self._memoize_outputs` attrib in init...?  I *really* want us to avoid code complexity that isn't strictly needed... it is going to *kill* us at this stage.

Regardless, if we move to only using getters / setters a la Java, I think we should be consistent throughout, and I don't think we're anywhere near there...

Reply:
☠️

Of course we're not going towards setters and getters, but I do see people preferring this type of mode setting (e.g. `torch.nn.Module.train(...)`) for things that actual users interact with (not just other classes). Open to just making `memoize` a public attrib for now.

Reply:
Although `train(...)` sets train mode recursively, so that's a bad example. Ok let's make it a public attrib